File: Android/app/src/main/java/com/example/reativos/oscar/Command.java
package com.example.reativos.oscar;

import java.util.List;
import java.util.ArrayList;

/**
 * Created by calmattoso on 6/18/16.
 */
public class Command {
    public List<Command> children = new ArrayList<>();
    public String type;
    public int param;

    Command(String type, int param) {
        this.type = type;
        this.param = param;
    }

    Command(String type) {
        this(type, -1);
    }

    public boolean addCommand(Command c) {
        if (type.equals("loop") && !c.type.equals("loop")) {
            children.add(c);
            return true;
        }
        return false;
    }

    public boolean removeCommand(int index) {
        if (index > children.size())
            return false;
        children.remove(index);
        return true;
    }

    public String serialize() {
        String str = this.type;

        if (type.equals("loop")) {
            // tratar dos filhos
            for (Command c : children) {
                str += c.toString();
            }
        }
        else if (param != -1) {
            str += Integer.toString(param);
        }

        return str;
    }

    @Override
    public String toString() {
        String str = this.type + " ";

        if (type.equals("move")) {
            if (param == 1) str += "Forward";
            else if (param == 0) str += "Backward";
        }
        else if (type.equals("turn")) {
            if (param == 1) str += "Right";
            else if (param == 0) str += "Left";
        }
        else if (type.contains("led")) {
            if (str.charAt(3) == 'r')
                str = "red led ";
            else if (str.charAt(3) == 'b')
                str = "blue led ";

            if (param == 1) str += "ON";
            else if (param == 0) str += "OFF";
        }
        else if (type.equals("buzz") || type.equals("wait"))
            str += Integer.toString(param) + "ms";


        return str;
    }
}

File: Oscar/Command.swift
//
//  Command.swift
//  Oscar
//
//  Created by Michelle Valente on 29/07/16.
//  Copyright © 2016 Michelle Valente. All rights reserved.
//

import UIKit

class Command: NSObject {
    
    var children = [Command]()
    var type = String()
    var param = Int()
    
    init(type:String, param: Int) {
        self.type = type
        self.param = param
    }
    
    convenience init(type:String) {
     self.init(type: type, param: -1)
    }
    
    func addCommand(c : Command) -> Bool {
        if (type == "loop" && !(c.type == "loop")) {
            children.append(c)
            return true
        }
        return false;
    }
    
    func removeCommand(index:Int) -> Bool {
        if (index > children.count){
            return false;
        }
        children.removeAtIndex(index);
        return true;
    }
    
    
    func serialize() -> String{
        var str = self.type
        if(type == "loop")
        {

            for (_,c) in children.enumerate() {
                str += c.toString()
            }
        }
        else if (param != -1)
        {
            str += String(param)
        }
        
        return str
    }
    
    func toString() -> String {
        var str = self.type + " "
        
        if(type == "move"){
            if(param == 1)
            {
                str += "forward"
            }
            else if(param == 0)
            {
                str += "backward"
            }
        }
        else if(type == "turn")
        {
            if(param == 1)
            {
                str += "right"
            }
            else if(param == 0)
            {
                str += "left"
            }
        }
        else if(type.lowercaseString.rangeOfString("led") != nil)
        {
            if(type == "ledr")
            {
                str = "red led"
            }
            else if(type == "ledb")
            {
                str = "blue led"
            }
            
            str += " "
            
            if(param == 1)
            {
                str += "on"
            }
            else if(param == 0)
            {
                str += "off"
            }
        }
        else if(type == "buzz" || type == "wait")
        {
            str += String(param) + "ms"
        }
        
        return str
        
    }
}
